File: shared/kmp/src/commonMain/kotlin/lk/mageride/shared/mqtt/MqttTopics.kt
package lk.mageride.shared.mqtt

import lk.mageride.shared.data.models.Ulid

// The EMQX topic tree, exactly as `backend/contracts/realtime/mqtt-topics.md` §1 prints it
// (ADD §7.2, D6' §3.1).
//
// THE TOPIC IS NOT WHERE AUTHORISATION LIVES. EMQX binds `{vehicleId}` from the device's JWT or
// X.509 claim, so a device physically cannot publish under another vehicle's topic. Building a
// topic for a vehicle the credential does not cover produces a rejected publish, not a leak —
// which is why these builders take the id and never a free-form string.

/** MQTT delivery guarantee. The position plane is QoS 1 throughout; only diagnostics is QoS 0. */
public enum class MqttQos(public val level: Int) {
    /** Fire and forget. */
    AT_MOST_ONCE(0),

    /** At least once — the broker redelivers until it is acknowledged. */
    AT_LEAST_ONCE(1),

    /** Exactly once. Not used by MageRide: the `seq` watermark already makes replay idempotent. */
    EXACTLY_ONCE(2),
}

/** Who publishes on a topic, and who consumes it. */
public enum class MqttDirection {
    /** The driver app or a tracker publishes; the broker fans it in. */
    DEVICE_TO_BROKER,

    /** The platform publishes; the device subscribes. */
    BROKER_TO_DEVICE,

    /** The platform publishes; a backend consumer or an operator subscribes. */
    BROKER_TO_CONSUMER,
}

/**
 * A branch of the topic tree with the QoS and retain flag the contract fixes for it.
 *
 * Retain is not a detail. `status` is retained so a consumer that subscribes after a device went
 * offline still learns it is offline; `pos/live` retains the last sample so a fresh subscriber
 * gets a position immediately instead of waiting a cadence tick. `pos/replay` must **not** be
 * retained — a retained backlog entry would be re-delivered as though it were current.
 *
 * @property qos Delivery guarantee for this branch.
 * @property retain Whether the broker keeps the last message.
 * @property direction Who publishes.
 */
public enum class MqttTopicKind(
    public val qos: MqttQos,
    public val retain: Boolean,
    public val direction: MqttDirection,
) {
    /** `veh/{vehicleId}/pos/live` — the hot path. */
    POSITION_LIVE(MqttQos.AT_LEAST_ONCE, retain = true, direction = MqttDirection.DEVICE_TO_BROKER),

    /** `veh/{vehicleId}/pos/replay` — offline backlog, rate-limited and never retained (R-17). */
    POSITION_REPLAY(MqttQos.AT_LEAST_ONCE, retain = false, direction = MqttDirection.DEVICE_TO_BROKER),

    /** `veh/{vehicleId}/cmd` — downlink; the device subscribes to its own only. */
    COMMAND(MqttQos.AT_LEAST_ONCE, retain = false, direction = MqttDirection.BROKER_TO_DEVICE),

    /** `veh/{vehicleId}/status` — `online` / `offline`, the LWT topic (R-15, T-04). */
    STATUS(MqttQos.AT_LEAST_ONCE, retain = true, direction = MqttDirection.BROKER_TO_CONSUMER),

    /** `fleet/{operatorId}/+/pos/live` — operator-scoped wildcard subscription. */
    FLEET_LIVE(MqttQos.AT_LEAST_ONCE, retain = false, direction = MqttDirection.BROKER_TO_CONSUMER),

    /** `sys/diag/{vehicleId}` — device diagnostics, QoS 0. */
    DIAGNOSTICS(MqttQos.AT_MOST_ONCE, retain = false, direction = MqttDirection.DEVICE_TO_BROKER),
}

/** A parsed topic: which branch, and whose. */
public data class MqttTopicRef(
    public val kind: MqttTopicKind,
    public val vehicleId: Ulid? = null,
    public val operatorId: Ulid? = null,
)

/** Builders and a parser for the topic tree. */
public object MqttTopics {

    private const val VEHICLE_ROOT = "veh"
    private const val FLEET_ROOT = "fleet"
    private const val DIAG_ROOT = "sys"

    /** Live GPS samples, `veh/{vehicleId}/pos/live`. */
    public fun positionLive(vehicleId: Ulid): String = "$VEHICLE_ROOT/$vehicleId/pos/live"

    /** Offline-buffered backlog, `veh/{vehicleId}/pos/replay`. */
    public fun positionReplay(vehicleId: Ulid): String = "$VEHICLE_ROOT/$vehicleId/pos/replay"

    /** Downlink commands, `veh/{vehicleId}/cmd` — the cadence hint arrives here (R-07). */
    public fun command(vehicleId: Ulid): String = "$VEHICLE_ROOT/$vehicleId/cmd"

    /** Presence, `veh/{vehicleId}/status` — the last-will topic. */
    public fun status(vehicleId: Ulid): String = "$VEHICLE_ROOT/$vehicleId/status"

    /** Device diagnostics, `sys/diag/{vehicleId}`. */
    public fun diagnostics(vehicleId: Ulid): String = "$DIAG_ROOT/diag/$vehicleId"

    /** An operator's whole fleet, `fleet/{operatorId}/+/pos/live`. Consumers only. */
    public fun fleetLive(operatorId: Ulid): String = "$FLEET_ROOT/$operatorId/+/pos/live"

    /**
     * The four topics a driver app touches for one vehicle.
     *
     * `cmd` is the only one it subscribes to; the other three it publishes (or wills) on.
     */
    public fun forVehicle(vehicleId: Ulid): Map<MqttTopicKind, String> = mapOf(
        MqttTopicKind.POSITION_LIVE to positionLive(vehicleId),
        MqttTopicKind.POSITION_REPLAY to positionReplay(vehicleId),
        MqttTopicKind.COMMAND to command(vehicleId),
        MqttTopicKind.STATUS to status(vehicleId),
        MqttTopicKind.DIAGNOSTICS to diagnostics(vehicleId),
    )

    /**
     * Reads a concrete topic back into [MqttTopicRef], or `null` if it is not one of ours.
     *
     * Used on the receive side: a client subscribed to more than one topic has to know which
     * branch a delivery came from before it can pick a codec.
     */
    @Suppress("ReturnCount")
    public fun parse(topic: String): MqttTopicRef? {
        val parts = topic.split('/')
        if (parts.size < MIN_SEGMENTS) return null
        return when (parts[0]) {
            VEHICLE_ROOT -> parseVehicle(parts)

            FLEET_ROOT -> parseFleet(parts)

            DIAG_ROOT -> if (parts.size == DIAG_SEGMENTS && parts[1] == "diag") {
                MqttTopicRef(MqttTopicKind.DIAGNOSTICS, vehicleId = parts[2])
            } else {
                null
            }

            else -> null
        }
    }

    /** `veh/{id}/cmd` is the shortest topic in the tree. */
    private const val MIN_SEGMENTS = 3

    /** `sys/diag/{vehicleId}`. */
    private const val DIAG_SEGMENTS = 3

    /** `fleet/{operatorId}/{vehicleId}/pos/live`. */
    private const val FLEET_SEGMENTS = 5

    /** How much of a fleet topic comes before `pos/live`. */
    private const val FLEET_PREFIX_SEGMENTS = 3

    private fun parseVehicle(parts: List<String>): MqttTopicRef? {
        val vehicleId = parts[1]
        val tail = parts.drop(2)
        val kind = when (tail) {
            listOf("pos", "live") -> MqttTopicKind.POSITION_LIVE
            listOf("pos", "replay") -> MqttTopicKind.POSITION_REPLAY
            listOf("cmd") -> MqttTopicKind.COMMAND
            listOf("status") -> MqttTopicKind.STATUS
            else -> return null
        }
        return MqttTopicRef(kind, vehicleId = vehicleId)
    }

    private fun parseFleet(parts: List<String>): MqttTopicRef? {
        if (parts.size != FLEET_SEGMENTS || parts.drop(FLEET_PREFIX_SEGMENTS) != listOf("pos", "live")) return null
        return MqttTopicRef(
            MqttTopicKind.FLEET_LIVE,
            vehicleId = parts[2].takeUnless { it == "+" },
            operatorId = parts[1],
        )
    }
}

File: backend/src/MageRide.Shared/Mqtt/MqttTopics.cs
using System.Diagnostics.CodeAnalysis;

namespace MageRide.Shared.Mqtt;

/// <summary>Which branch of the EMQX topic tree a topic belongs to.</summary>
public enum MqttTopicKind
{
    /// <summary><c>veh/{vehicleId}/pos/live</c> — the hot path. QoS 1, retain last.</summary>
    PositionLive,

    /// <summary><c>veh/{vehicleId}/pos/replay</c> — offline backlog, rate-limited, never retained (R-17).</summary>
    PositionReplay,

    /// <summary><c>veh/{vehicleId}/cmd</c> — downlink; the device subscribes to its own only.</summary>
    Command,

    /// <summary><c>veh/{vehicleId}/status</c> — <c>online</c>/<c>offline</c>, the LWT topic (R-15, T-04).</summary>
    Status,

    /// <summary><c>sys/diag/{vehicleId}</c> — device diagnostics, QoS 0.</summary>
    Diagnostics,
}

/// <summary>A parsed topic: which branch, and whose.</summary>
public sealed record MqttTopicRef(MqttTopicKind Kind, Guid VehicleId);

/// <summary>
/// The EMQX topic tree, exactly as <c>backend/contracts/realtime/mqtt-topics.md</c> §1 prints it
/// (ADD §7.2, D6' §3.1) — the server half of the KMP module's <c>MqttTopics</c>.
/// </summary>
/// <remarks>
/// <para>
/// <b>The topic is not where authorisation lives.</b> EMQX binds <c>{vehicleId}</c> from the
/// device's JWT claim (<c>infra/deploy/emqx/emqx.conf</c>'s <c>verify_claims</c>) and
/// <c>acl.conf</c> writes every device rule in terms of <c>${username}</c>, so a device physically
/// cannot publish under another vehicle's topic. Building a topic for a vehicle the credential does
/// not cover produces a rejected publish and a disconnect, not a leak — which is why these builders
/// take a <see cref="Guid"/> and never a free-form string.
/// </para>
/// <para>
/// The bridge consumes through <b>shared subscriptions</b>: <c>$share/{group}/veh/+/pos/live</c>.
/// EMQX dispatches each message to exactly one member of the group, which is what lets
/// mqtt-bridge-svc scale horizontally without duplicate ingest into <c>telemetry.raw</c> (E-08).
/// </para>
/// </remarks>
public static class MqttTopics
{
    private const string VehicleRoot = "veh";
    private const string DiagRoot = "sys";

    /// <summary>
    /// The shared-subscription group mqtt-bridge-svc's live consumers join (ADD §7.3, D6' §3.3).
    /// </summary>
    public const string LiveShareGroup = "posGroup";

    /// <summary>
    /// The parallel group for the replay stream. <b>Separate on purpose</b>: R-09 splits live from
    /// replay so a fleet reconnecting after a regional outage cannot drown live samples with its
    /// backlog, and the two groups are what makes the priorities independent.
    /// </summary>
    public const string ReplayShareGroup = "posReplayGroup";

    /// <summary>Live GPS samples, <c>veh/{vehicleId}/pos/live</c>.</summary>
    public static string PositionLive(Guid vehicleId) => $"{VehicleRoot}/{vehicleId}/pos/live";

    /// <summary>Offline-buffered backlog, <c>veh/{vehicleId}/pos/replay</c>.</summary>
    public static string PositionReplay(Guid vehicleId) => $"{VehicleRoot}/{vehicleId}/pos/replay";

    /// <summary>Downlink commands, <c>veh/{vehicleId}/cmd</c> — the cadence hint arrives here (R-07).</summary>
    public static string Command(Guid vehicleId) => $"{VehicleRoot}/{vehicleId}/cmd";

    /// <summary>Presence, <c>veh/{vehicleId}/status</c> — the last-will topic.</summary>
    public static string Status(Guid vehicleId) => $"{VehicleRoot}/{vehicleId}/status";

    /// <summary>Device diagnostics, <c>sys/diag/{vehicleId}</c>.</summary>
    public static string Diagnostics(Guid vehicleId) => $"{DiagRoot}/diag/{vehicleId}";

    /// <summary>The wildcard every vehicle's live stream matches: <c>veh/+/pos/live</c>.</summary>
    public const string AllPositionsLive = "veh/+/pos/live";

    /// <summary>The wildcard every vehicle's replay stream matches: <c>veh/+/pos/replay</c>.</summary>
    public const string AllPositionsReplay = "veh/+/pos/replay";

    /// <summary>
    /// A load-balanced filter: <c>$share/{group}/{filter}</c>. Every replica in
    /// <paramref name="group"/> gets a share of the messages, none gets a copy (E-08).
    /// </summary>
    public static string Shared(string group, string filter)
    {
        ArgumentException.ThrowIfNullOrWhiteSpace(group);
        ArgumentException.ThrowIfNullOrWhiteSpace(filter);

        return $"$share/{group}/{filter}";
    }

    /// <summary><c>$share/posGroup/veh/+/pos/live</c> — the bridge's live subscription.</summary>
    public static string SharedPositionLive(string group = LiveShareGroup) => Shared(group, AllPositionsLive);

    /// <summary><c>$share/posReplayGroup/veh/+/pos/replay</c> — the bridge's backlog subscription.</summary>
    public static string SharedPositionReplay(string group = ReplayShareGroup) => Shared(group, AllPositionsReplay);

    /// <summary>
    /// Reads a concrete topic back, or <see langword="false"/> when it is not one of ours.
    /// </summary>
    /// <remarks>
    /// Used on the receive side. A shared subscription delivers the <i>concrete</i> topic, not the
    /// filter, so this is how the bridge learns which vehicle a payload belongs to — the partition
    /// key for <c>telemetry.raw</c> comes from the topic, never from the payload, because the topic
    /// is the part EMQX authenticated.
    /// </remarks>
    public static bool TryParse(string? topic, [NotNullWhen(true)] out MqttTopicRef? reference)
    {
        reference = null;

        if (string.IsNullOrWhiteSpace(topic))
        {
            return false;
        }

        var parts = topic.Split('/');

        if (parts.Length == 3 && parts[0] == DiagRoot && parts[1] == "diag" && Guid.TryParse(parts[2], out var diagId))
        {
            reference = new MqttTopicRef(MqttTopicKind.Diagnostics, diagId);
            return true;
        }

        if (parts.Length < 3 || parts[0] != VehicleRoot || !Guid.TryParse(parts[1], out var vehicleId))
        {
            return false;
        }

        MqttTopicKind? kind = (parts.Length, parts[2]) switch
        {
            (4, "pos") when parts[3] == "live" => MqttTopicKind.PositionLive,
            (4, "pos") when parts[3] == "replay" => MqttTopicKind.PositionReplay,
            (3, "cmd") => MqttTopicKind.Command,
            (3, "status") => MqttTopicKind.Status,
            _ => null,
        };

        if (kind is null)
        {
            return false;
        }

        reference = new MqttTopicRef(kind.Value, vehicleId);
        return true;
    }
}

/// <summary>Presence on <c>veh/{vehicleId}/status</c> — the retained LWT payload (R-15, T-04).</summary>
public static class VehicleStatus
{
    /// <summary>Published by the device after a successful CONNECT.</summary>
    public const string Online = "online";

    /// <summary>The broker's last will, or the TCP adapter's emulation of it on half-close (T-04).</summary>
    public const string Offline = "offline";
}
